salt/modules/zypper.py

```diff
@@ -58,65 +61,223 @@ def __virtual__():
     return __virtualname__
 
 
-def _zypper(*opts):
-    '''
-    Return zypper command with default options as a list.
-
-    opts
-        additional options for zypper command
-
-    '''
-    cmd = ['zypper', '--non-interactive']
-    cmd.extend(opts)
-
-    return cmd
-
-
-def _is_zypper_error(retcode):
-    '''
-    Return True in case the exist code indicate a zypper errror.
-    Otherwise False
-    '''
-    # see man zypper for existing exit codes
-    return int(retcode) not in [0, 100, 101, 102, 103]
+class _Zypper(object):
+    '''
+    Zypper parallel caller.
+    Validates the result and either raises an exception or reports an error.
+    Allows serial zypper calls (first came, first won).
+    '''
+
+    SUCCESS_EXIT_CODES = [0, 100, 101, 102, 103]
+    LOCK_EXIT_CODE = 7
+    XML_DIRECTIVES = ['-x', '--xmlout']
+    ZYPPER_LOCK = '/run/zypp.pid'
+    TAG_RELEASED = 'zypper/released'
+    TAG_BLOCKED = 'zypper/blocked'
+
+    def __init__(self, no_refresh=True):
+        '''
+        Constructor
+        :param no_refresh:
+        '''
+        self.__called = False
+        self._reset()
+
+    def _reset(self):
+        '''
+        Resets values of the call setup.
+
+        :return:
+        '''
+        self.__cmd = ['zypper', '--non-interactive']
+        self.__exit_code = 0
+        self.__call_result = dict()
+        self.__error_msg = ''
+        self.__env = {'SALT_RUNNING': "1"}  # Subject to change
+
+        # Call config
+        self.__xml = False
+        self.__no_lock = False
+        self.__no_raise = False
+        self.__refresh = False
+
+    def __getattr__(self, item):
+        '''
+        Call configurator.
+
+        :param item:
+        :return:
+        '''
+        # Reset after the call
+        if self.__called:
+            self._reset()
+            self.__called = False
+
+        if item == 'xml':
+            self.__xml = True
+        elif item == 'nolock':
+            self.__no_lock = True
+        elif item == 'noraise':
+            self.__no_raise = True
+        elif item == 'refreshable':
+            self.__refresh = True
```

Comment:
We need to prevent the use of "refreshable" together with "nolock".
If we allow refresh, than we are not anymore in a read-only mode.